### PR TITLE
Add more docs - mostly warnings - to std::mem::transmute

### PR DESCRIPTION
**Pull request by strega-nil**

### CODE REVIEW

**Comment by ubsan** [comment]
Perhaps I should also add an example for casting to `usize`.

**Comment by arielb1** [comment]
`transmute` can also be used to transform a `Vec<Option<&u32>>` to a `Vec<&u32>`.

**Comment by ubsan** [comment]
@arielb1 Ah, good for the alternatives section!

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -278,17 +278,157 @@ extern "rust-intrinsic" {
     /// Moves a value out of scope without running drop glue.
     pub fn forget<T>(_: T) -> ();
 
-    /// Unsafely transforms a value of one type into a value of another type.
+    /// Reinterprets the bits of a value of one type as another type. Both types
```
Comment (ollie27) [inline src/libcore/intrinsics.rs]: The first paragraph needs to be a single sentence as it's used as a short summary on the module page and search results.

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -278,17 +278,157 @@ extern "rust-intrinsic" {
...
+    /// must have the same size. Neither the original, nor the result, may be an
+    /// [invalid value]
+    /// (https://doc.rust-lang.org/nomicon/meet-safe-and-unsafe.html).
```
Comment (ollie27) [inline src/libcore/intrinsics.rs]: This link should be `../../nomicon/meet-safe-and-unsafe.html` so it works in the offline docs.

File: src/libcore/intrinsics.rs [hunk 3]
```diff
@@ -278,17 +278,157 @@ extern "rust-intrinsic" {
...
+    ///     mem::forget(t);
+    ///     u
+    /// }
+    /// ```
+    ///
+    /// `transmute` is incredibly unsafe. There are a vast number of ways to
+    /// cause undefined behavior with this function. `transmute` should be
+    /// the absolute last resort.
+    ///
+    /// The [nomicon](https://doc.rust-lang.org/nomicon/transmutes.html) has
```
Comment (ollie27) [inline src/libcore/intrinsics.rs]: Same as above.

Comment (ubsan) [inline src/libcore/intrinsics.rs]: @ollie27 thank you for these :)

File: src/libcore/intrinsics.rs [hunk 4]
```diff
@@ -278,17 +278,157 @@ extern "rust-intrinsic" {
...
+    ///         assert!(index < len);
+    ///         let slice2 = mem::transmute::<&mut [T], &mut [T]>(slice);
+    ///         (slice[0..index], slice2[index..len])
+    ///     }
+    ///     // Again, use `as` and reborrowing
+    ///     fn split_at_mut_casts<T>(slice: &mut [T], index: usize)
+    ///                              -> (&mut [T], &mut [T]) {
+    ///         let len = slice.len();
+    ///         assert!(index < len);
+    ///         let slice2 = &mut *(slice as *mut [T]); // actually typesafe!
```
Comment (ollie27) [inline src/libcore/intrinsics.rs]: `slice` and `slice2` are aliased here so is this really safe?

This isn't how [`split_at_mut`](https://github.com/rust-lang/rust/blob/master/src/libcore/slice.rs#L339-L349) is actually implemented so I'm not sure this is a good example and the advice should be just to use the method in `std`.

**Comment by ollie27** [comment]
A lot of these examples aren't compiling, check the travis output and try them locally. I suggest you put each example in it's own code block so they run individually and some examples like the `libc` one need to be marked `ignore` as that won't even compile.

**Comment by ubsan** [comment]
@ollie27 Yeah, I've obviously never written documentation like this before xP

**Comment by ubsan** [comment]
@ollie27 Okay, I can't actually try it locally due to LLDB failures. I'm... not willing to continue to try this. If someone else wants to pick this up, they can, but goddamnit, I've spent a fucking *day* trying to get rustc to compile. This is ridiculous.

**Comment by ollie27** [comment]
You shouldn't need to compile `rustc` to try these. If you have a working `rustdoc` you can copy this doc block into it's own file and run `rustdoc --test` on that. You can also use https://play.rust-lang.org/.

**Comment by steveklabnik** [comment]
> I suggest you put each example in it's own code block so they run individually and some examples like the libc one need to be marked ignore as that won't even compile.

Yes please to both of these!

File: src/libcore/intrinsics.rs [hunk 5]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// more complete documentation. Read it before using `transmute`.
+    ///
+    /// # Alternatives
+    ///
+    /// There are very few good cases for `transmute`. Most can be achieved
+    /// through other means. Some more or less common uses, and a better way,
+    /// are as follows:
+    ///
+    /// Turning a pointer into a `usize`:
+    /// ```
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: can you put a blank line between the triple backticks and the descriptions? This applies throughout the file.

File: src/libcore/intrinsics.rs [hunk 6]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// let ptr_num_transmute = mem::transmute::<&i32, usize>(ptr);
+    /// // Use `as` casts instead
+    /// let ptr_num_cast = ptr as *const i32 as usize;
+    /// ```
+    ///
+    /// Turning a `*mut T` into an `&mut T`:
+    /// ```
+    /// let ptr: *mut i32 = &mut 0;
+    /// let ref_transmuted = mem::transmute::<*mut i32, &mut i32>(ptr);
+    /// // Use reborrows
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: nit: "Use a reborrow"

File: src/libcore/intrinsics.rs [hunk 7]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// // literal
+    /// assert_eq!(b"Rust", [82, 117, 116, 116]);
+    /// ```
+    ///
+    /// Turning a `Vec<&T>` into a `Vec<Option<&T>>`:
+    /// ```
+    /// let store = [0, 1, 2, 3];
+    /// let v_orig = store.iter().collect::<Vec<&i32>>();
+    /// // Using transmute: this is Undefined Behavior, and a bad idea
+    /// // However, it is no-copy
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: needs a period

File: src/libcore/intrinsics.rs [hunk 8]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// let v_transmuted = mem::transmute::<Vec<&i32>, Vec<Option<&i32>>>(
+    ///     v_orig.clone());
+    /// // This is the suggested, safe way
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: and here, and the line below

File: src/libcore/intrinsics.rs [hunk 9]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// There are valid uses of transmute, though they are few and far between.
+    ///
+    /// Getting the bitpattern of a floating point type:
+    /// ```
+    /// let bitpattern = std::mem::transmute::<f32, u32>(1.0);
+    /// assert_eq!(bitpattern, 0x3F800000);
+    /// ```
+    ///
+    /// Turning a pointer into a function pointer (this isn't guaranteed to
+    /// work in Rust, although, for example, Linux does make this guarantee):
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: If it's not guaranteed, why are we recommending it?

File: src/libcore/intrinsics.rs [hunk 10]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// }
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn transmute<T, U>(e: T) -> U;
 
+    /// Gives the address for the return value of the enclosing function.
+    ///
+    /// Using this intrinsic in a function that does not use an out pointer
+    /// will trigger a compiler error.
+    pub fn return_address() -> *const u8;
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: was this change intentional?

**Comment by steveklabnik** [comment]
Thanks @ubsan, mostly nits about formatting, etc. I'm glad to get more warnings about `transmute` in the docs, I agree that it's still used too often today.

@rust-lang/lang @rust-lang/libs I'd like you to weigh in on these recommendations.

File: src/libcore/intrinsics.rs (hunk 9 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: Because it's useful on some platforms (like POSIX and Windows :P)

File: src/libcore/intrinsics.rs (hunk 10 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: Ermm... no

File: src/libcore/intrinsics.rs [hunk 11]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// // Use reborrows
+    /// let ref_casted = &mut *ptr;
+    /// ```
+    ///
+    /// Turning an `&mut T` into an `&mut U`:
+    /// ```
+    /// let ptr = &mut 0;
+    /// let val_transmuted = mem::transmute::<&mut i32, &mut u32>(ptr);
+    /// // Now let's put together `as` and reborrowing
+    /// let val_casts = &mut *(ptr as *mut i32 as *mut u32);
```
Comment (sfackler) [inline src/libcore/intrinsics.rs]: It is not clear to me why this is preferable.

File: src/libcore/intrinsics.rs [hunk 12]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    ///          slice::from_raw_parts_mut(ptr.offset(mid as isize), len - mid))
+    ///     }
+    /// }
+    /// ```
     ///
     /// # Examples
     ///
+    /// There are valid uses of transmute, though they are few and far between.
+    ///
+    /// Getting the bitpattern of a floating point type:
```
Comment (sfackler) [inline src/libcore/intrinsics.rs]: Why is this a valid use case and not up above under the "just type pun with raw pointers" examples?

**Comment by sfackler** [comment]
As far as I can tell, these docs are basically saying "transmute is a terrible, rotten, no-good function. Here are 5 examples where you shouldn't use it. Here are 3 examples where you should use it." Coming out of it, I have no intuition as to why the bad examples are bad and the good examples are good. In particular, many of the "don't use transmute for X" examples replace the transmute with casts of raw pointers. As far as I can tell, doing that still has *exactly the same* problems that transmute does with respect to trap values etc. In addition, casting through raw pointers removes the check that both types have the same size that transmute does provide.

File: src/libcore/intrinsics.rs (hunk 9 above)
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: Then we should describe it correctly.

**Comment by ubsan** [comment]
@sfackler Uh... what?

File: src/libcore/intrinsics.rs (hunk 12 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: Because the above aren't type punning with raw pointers.

File: src/libcore/intrinsics.rs (hunk 11 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: `transmute` is too powerful.

File: src/libcore/intrinsics.rs (hunk 12 above)
Comment (sfackler) [inline src/libcore/intrinsics.rs]: So is `*(&1.0 as *const f32 as *const u32)` a Bad Thing, or just an alternative to this?

Comment (ubsan) [inline src/libcore/intrinsics.rs]: a Bad Thing.

File: src/libcore/intrinsics.rs (hunk 11 above)
Comment (sfackler) [inline src/libcore/intrinsics.rs]: That is not stated anywhere in these docs. I also am still not clear on exactly what that means, or why `transmute`'s ability to do other things is a problem when someone wants to turn a `&mut i32` to an `&mut u32`.

`&mut *(ptr as *mut i32 as *mut u32)` seems super symbol-soupy to me, whereas `transmute::<&mut i32, &mut u32>(ptr)` makes it very clear what's going on. Nothing in these docs give me any indication whatsoever as to *why* I should go with the soup, just that I *should*.

File: src/libcore/intrinsics.rs (hunk 12 above)
Comment (sfackler) [inline src/libcore/intrinsics.rs]: I know that we're not dealing with pointers here, but again, the docs do not explain anywhere as to why that's as important of a distinction as it apparently is.

File: src/libcore/intrinsics.rs (hunk 9 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: How do you mean?

File: src/libcore/intrinsics.rs (hunk 12 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: > `transmute` is incredibly unsafe. There are a vast number of ways to
> cause undefined behavior with this function. `transmute` should be
> the absolute last resort.

I'm not sure what else you want. I also link to the nomicon page on it. `transmute` is generally agreed to be a Bad Idea. These aren't meant to be the be-all end-all docs; these are meant to stop you in your tracks, and show you what you should be using instead.

File: src/libcore/intrinsics.rs (hunk 11 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: No, it doesn't. It doesn't make it clear at all that it's a *bitcast* between `&mut i32` and `&mut u32`. A *memcpy* is happening.

Comment (sfackler) [inline src/libcore/intrinsics.rs]: It is clear to everyone that knows what transmute is. If you don't know what transmute is it may not be clear, just as it may not be clear what the casting implementation did if you didn't know what raw pointers were in Rust, or that `&mut *` does not actually involve a read of the underlying value and a copy into a new storage location.

File: src/libcore/intrinsics.rs (hunk 12 above)
Comment (sfackler) [inline src/libcore/intrinsics.rs]: If all they're meant to do is stop you in your tracks we don't need 400 lines of examples. An `<h1><blink><marquee>DON'T USE THIS UNLESS YOU KNOW WHAT YOU'RE DOING</marquee></blink></h1>` would serve that purpose.

File: src/libcore/intrinsics.rs [hunk 13]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    /// // assuming that T and U are the same size
+    /// unsafe fn transmute<T, U>(t: T) -> U {
+    ///     let mut u: U = mem::uninitialized();
+    ///     ptr::copy_nonoverlapping(&t as *const T as *const u8,
+    ///                              &mut u as *mut U as *mut u8,
+    ///                              mem::size_of::<T>());
+    ///     mem::forget(t);
+    ///     u
+    /// }
+    /// ```
```
Comment (nrc) [inline src/libcore/intrinsics.rs]: I feel that this would be more useful to most users if explained intuitively, rather than by giving code.

File: src/libcore/intrinsics.rs [hunk 14]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    ///
+    /// `transmute` is incredibly unsafe. There are a vast number of ways to
+    /// cause undefined behavior with this function. `transmute` should be
+    /// the absolute last resort.
```
Comment (nrc) [inline src/libcore/intrinsics.rs]: I object to the tone here, I think it will serve only to annoy readers, rather than warn them off. Rather than an implicit argument from authority, it would be better to say *why* transmute is unsafe and how it can cause UB and why that is bad.

File: src/libcore/intrinsics.rs [hunk 15]
```diff
@@ -278,21 +278,199 @@ extern "rust-intrinsic" {
...
+    ///
+    /// The [nomicon](https://doc.rust-lang.org/nomicon/transmutes.html) has
+    /// more complete documentation. Read it before using `transmute`.
```
Comment (nrc) [inline src/libcore/intrinsics.rs]: "Read it before using `transmute`" is unnecessary and patronising

**Comment by nrc** [comment]
I agree on the technicalities here, but agree with sfackler that it would be better to give some intuition of where transmute should and should not be used, rather than (or as well as) specific examples. E.g., "where possible, it is better to use casts rather than transmutation, for example...".

File: src/libcore/intrinsics.rs (hunk 15 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: I don't think it's unnecessary. You *should* read these docs before using transmute, transmute is one of the most Undefined Behavior-causing, misused parts of Rust. It's an advanced part of Rust.

File: src/libcore/intrinsics.rs (hunk 14 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: Perhaps. However, we already have the nomicon, and that is the point of it.

File: src/libcore/intrinsics.rs (hunk 15 above)
Comment (nrc) [inline src/libcore/intrinsics.rs]: I agree, but the fact that there is a reference to it here in the docs implies that you should read it, you don't need to spell it out too. I think it will make readers less likely to actually read the reference because they will be irritated.

File: src/libcore/intrinsics.rs (hunk 13 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: How does one "explain intuitively" this concept. For comparison, "Unsafely transforms" is what the previous docs had. Many people think it does a `reinterpret_cast`. I think it's necessary to have a real code example, in order to tell one what's actually happening.

**Comment by ubsan** [comment]
@nrc The idea is to give people an alternative. Beginners don't use `transmute` because they want to. They use `transmute` because they don't think they have any other choice. This is intended to show common patterns where people use `transmute` and they don't need to, and what they can use instead.

There are four intentions to this documentation:

First, the `memcpy` example. This clears up a common misconception, which is that `transmute<T, U>(t: T)` is equivalent to `reinterpret_cast<U*>(&(t: T))`.

Second, the "`transmute` is incredibly unsafe... with this function". You should not use this function if you have any alternatives.

Third, the Alternatives section: what you can use instead.

Fourth, the Example section: what `transmute` *is* actually good for.

File: src/libcore/intrinsics.rs (hunk 14 above)
Comment (nrc) [inline src/libcore/intrinsics.rs]: I think it would be good to have a sentence here which briefly describes why transmute is unsafe, and hopefully inspires the reader to go to the nomicon to find out more. We can't expect that every reader will also read the nomicon.

> We want to keep away people however possible

I don't agree. We want to give users enough information to make an informed choice and to properly understand the trade-offs they are making with that choice.

Comment (ubsan) [inline src/libcore/intrinsics.rs]: I guess... I've just seen so many people use it poorly, making their Rust code less safe.

File: src/libcore/intrinsics.rs (hunk 13 above)
Comment (nrc) [inline src/libcore/intrinsics.rs]: Something like "`transmute` returns a bitwise copy of the input. Before returning, the input's memory is freed without calling any destructor or otherwise tidying up"

File: src/libcore/intrinsics.rs (hunk 12 above)
Comment (ubsan) [inline src/libcore/intrinsics.rs]: See https://github.com/rust-lang/rust/pull/34609#issuecomment-230621533

**Comment by ubsan** [comment]
I am so bad at these tools Dx

**Comment by alexandermerritt** [comment]
I am new to Rust, and I'd like to mention where I get lost and confused (as an aside, I think other newcomers should provide their inputs why they are using ```transmute``` and their thinking around it to help improve this situation). But I agree with @ubsan's sentiment that in many scenarios I am unaware of the tricks to apply ```as``` in the right way to avoid ```transmute```.

I understand ```transmute``` to Just Do It but that there are better methods for transformation. It would be very helpful for me to know what the underlying rules are as to why ```as``` works or doesn't allow something, or if there are other ways to acquire the end type I want that uses neither ```as``` nor ```transmute```; a philosophy of what ```as``` is for.

After a few ad-hoc attempts, I may give up and use transmute:

    let iref = &0;
    let p = iref as *const u8;
    // error: casting `&usize` as `*const u8` is invalid

or, as in the first example in this PR,

    let p = &0_usize as usize;
    // error: casting `&usize` as `usize` is invalid

"invalid" is not helpful and I would not think of using a chained ```as``` to acquire the resulting ```usize```-represented address. There is no intuition.

[The Book](https://doc.rust-lang.org/book/casting-between-types.html#explicit-coercions) has a list of what is valid, but I feel it could benefit from better organization. For example, not just be a list of what works.

Documentation for ```transmute``` should begin first by telling me when I may benefit greatly from using it, how it functions, etc. motivating its purpose, then e.g., "outside of these use-cases, you should use ```as``` or do something else" and avoid conveying use of ```transmute``` as the "forbidden tool". Then follow up with "here are common traps where you may feel the urge to use transmute, the reason being due to incomplete docs, or the way ```as``` works is currently not intuitive, compiler output not constructive, ..." etc.

Documentation for ```transmute``` may benefit from constructive documentation for ```as``` preceding it.

**Comment by ubsan** [comment]
@alexandermerritt thanks :)

File: src/libcore/intrinsics.rs [hunk 16]
```diff
@@ -278,17 +278,200 @@ extern "rust-intrinsic" {
...
+    ///                                              -> &'b mut R<'c> {
+    ///     std::mem::transmute::<&'b mut R<'static>, &'b mut R<'c>>(r)
+    /// }
+    /// ```
+    ///
+    /// # Alternatives
+    ///
+    /// However, many uses of `transmute` can be achieved through other means.
+    /// This is unfortunate because either `transmute` isn't guaranteed to work
+    /// in that case, and only does because of rustc's current implemenation;
```
Comment (alexandermerritt) [inline src/libcore/intrinsics.rs]: "implemenation"

File: src/libcore/intrinsics.rs [hunk 17]
```diff
@@ -278,17 +278,200 @@ extern "rust-intrinsic" {
...
+    /// or, more commonly, `transmute` is just too powerful. It can transform
+    /// any type into any other, with just the caveat that they're the same
+    /// size. Some more or less common uses, and a better way, are as follows:
```
Comment (alexandermerritt) [inline src/libcore/intrinsics.rs]: Below are common uses of `transmute` that can be replaced with safe applications of `as` instead:

File: src/libcore/intrinsics.rs [hunk 18]
```diff
@@ -278,17 +278,200 @@ extern "rust-intrinsic" {
...
+    ///
+    /// ```
+    /// let ptr = &0;
+    /// let ptr_num_transmute = unsafe {
+    ///     std::mem::transmute::<&i32, usize>(ptr)
+    /// };
+    /// // Use an `as` cast instead
+    /// let ptr_num_cast = ptr as *const i32 as usize;
+    /// ```
+    ///
```
Comment (alexandermerritt) [inline src/libcore/intrinsics.rs]: Note the chaining of `as` to "hop" between types. `as` is not transitive; it requires explicit conversions between each step. (or something, to give intuition as what in the world is going on here :-D )

File: src/libcore/intrinsics.rs [hunk 19]
```diff
@@ -278,17 +278,201 @@ extern "rust-intrinsic" {
...
-    /// Unsafely transforms a value of one type into a value of another type.
+    /// Reinterprets the bits of a value of one type as another type; both types
+    /// must have the same size. Neither the original, nor the result, may be an
+    /// [invalid value] (../../nomicon/meet-safe-and-unsafe.html).
     ///
-    /// Both types must have the same size.
+    /// `transmute` is semantically equivalent to a bitwise move of one type
+    /// into another. It copies the bits from the destination type into the
+    /// source type, then forgets the original. If you know C or C++, it's like
+    /// `memcpy` under the hood.
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: We usually don't refer to other languages in the docs. I would just say 

> then forgets the original, like `memcpy`.

File: src/libcore/intrinsics.rs [hunk 20]
```diff
@@ -278,17 +278,201 @@ extern "rust-intrinsic" {
...
+    ///
+    /// The [nomicon](../../nomicon/transmutes.html) has additional
+    /// documentation.
     ///
     /// # Examples
     ///
+    /// There are a few things that `transmute` is really useful for.
+    ///
+    /// Getting the bitpattern of a floating point type (or, more generally,
+    /// type punning, when T and U aren't pointers):
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: Do we have any docs at all on type punning? Would be cool to link if so.

(and `T` and `U` should be in backticks)

File: src/libcore/intrinsics.rs [hunk 21]
```diff
@@ -278,17 +278,201 @@ extern "rust-intrinsic" {
...
+    ///                                              -> &'b mut R<'c> {
+    ///     std::mem::transmute::<&'b mut R<'static>, &'b mut R<'c>>(r)
+    /// }
+    /// ```
+    ///
+    /// # Alternatives
+    ///
+    /// However, many uses of `transmute` can be achieved through other means.
+    /// `transmute` can transform
+    /// any type into any other, with just the caveat that they're the same
```
Comment (steveklabnik) [inline src/libcore/intrinsics.rs]: would you mind reflowing this

**Comment by steveklabnik** [comment]
@ubsan thanks for all the work on this, it's improved greatly :smile: 

I have some extremely small nits and then this is good to go, I think.

**Review by tamird — COMMENTED** [review]
(leaving a comment here in case the author and reviewers have opinions)

Why is using pointer-to-pointer `as` casts most sensible (or safer) than `transmute`? In 2025, involving pointers might trigger questions about provenance. Is that better?

This came up in [a discussion on the Rust-for-Linux mailing list](https://lore.kernel.org/all/CAJ-ks9ntTxBM=c5nUZWGv3MoRt-LveBchn-c1Xy-DGap7fLVRA@mail.gmail.com/), and I commented about it [another issue](https://github.com/rust-lang/rust/issues/34249#issuecomment-2755316610) as well.

There's also been some whiplash on this in std itself. `transmute` was replaced with this pattern in https://github.com/rust-lang/rust/commit/2633b85ab2c89822d2c227fc9e81c6ec1c0ed9b6, and then replaced with transmute again in https://github.com/rust-lang/rust/pull/75157/files.

I'd love to understand how this came to be the suggested pattern.

EDIT: @sfackler said it in https://github.com/rust-lang/rust/pull/34609#issuecomment-230559871. Shame this wasn't addressed.

**Comment by ChrisDenton** [comment]
Eight years is a long time in software. It's very possible everyone involved in this PR has moved on.

You'd be better off opening a new issue or asking on the [Zulip chat](https://rust-lang.zulipchat.com).

**Comment by tamird** [comment]
> Nine years is a long time in software. It's very possible everyone involved in this PR has moved on.
> 
> You'd be better off opening a new issue or asking on the [Zulip chat](https://rust-lang.zulipchat.com).

[#general > casting &amp;&#91;T&#93; to &#96;#&#91;repr(transparent)&#93;&#96; wrapper struct @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/casting.20.26.5BT.5D.20to.20.60.23.5Brepr.28transparent.29.5D.60.20wrapper.20struct/near/508539237) for a start.
